server/src/main/java/org/opensearch/index/store/remote/file/OnDemandVirtualFileSnapshotIndexInput.java

```diff
@@ -0,0 +1,133 @@
+     */
+    private final TransferManager transferManager;
+
+    /**
+     * FileInfo contains snapshot metadata references for this IndexInput
+     */
+    private final FileInfo fileInfo;
+
+    /**
+     * underlying lucene directory to open file
+     */
+    protected final FSDirectory directory;
+
+    /**
+     * file name
+     */
+    protected final String fileName;
+
+    public OnDemandVirtualFileSnapshotIndexInput(FileInfo fileInfo, FSDirectory directory, TransferManager transferManager) {
+        this(
+            "VirtualFileSnapshotIndexInput(path=\""
+                + directory.getDirectory().toString()
+                + "/"
+                + fileInfo.physicalName()
+                + "\", "
+                + "offset="
+                + 0
+                + ", length= "
+                + fileInfo.length()
+                + ")",
+            fileInfo,
+            directory,
+            transferManager,
+            fileInfo.physicalName(),
+            0L,
+            fileInfo.length(),
+            false
+        );
+    }
+
+    public OnDemandVirtualFileSnapshotIndexInput(
+        String resourceDescription,
+        FileInfo fileInfo,
+        FSDirectory directory,
+        TransferManager transferManager,
+        String fileName,
+        long offset,
+        long length,
+        boolean isClone
+    ) {
+        super(
+            OnDemandBlockIndexInput.builder()
+                .resourceDescription(resourceDescription)
+                .isClone(isClone)
+                .offset(offset)
+                .length(length)
+                .blockSizeShift(BLOCK_SIZE_SHIFT)
+        );
+        this.fileInfo = fileInfo;
+        this.directory = directory;
+        this.fileName = fileName;
+        this.transferManager = transferManager;
+    }
+
+    @Override
+    protected OnDemandVirtualFileSnapshotIndexInput buildSlice(String sliceDescription, long offset, long length) {
+        return new OnDemandVirtualFileSnapshotIndexInput(
+            sliceDescription,
+            this.fileInfo,
+            this.directory,
+            this.transferManager,
+            this.fileName,
+            offset,
+            length,
+            true
+        );
+    }
+
+    @Override
+    public OnDemandVirtualFileSnapshotIndexInput clone() {
+        OnDemandVirtualFileSnapshotIndexInput clone = buildSlice("clone", 0L, this.length);
+        // ensures that clones may be positioned at the same point as the blocked file they were cloned from
+        if (currentBlock != null) {
+            clone.currentBlock = currentBlock.clone();
+            clone.currentBlockId = currentBlockId;
+        }
+        return clone;
+    }
+
+    @Override
+    protected IndexInput fetchBlock(int blockId) throws IOException {
+        // will always have one block. this method will be used to ensure that virtual file is cached into local fs
+        final Path filePath = directory.getDirectory().resolve(fileName);
```

Comment:
Should we close the Directory on stale/bad file paths

Reply:
sorry again about this misleading comment. in this phase, there will be no file system caching so just changed the comment here. now for closing the directory, this directory will be singleton per shard per node and it's closing is handled in `RemoteSnapshotDirectory::close` not within the file here